File: java/examples/cc0mon-api-get-registry-images.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 17
//DEPS com.cc0mon:sdk:0.1.0
//DEPS com.fasterxml.jackson.core:jackson-databind:2.17.0

import com.cc0mon.sdk.Client;
import com.cc0mon.sdk.Errors;
import com.cc0mon.sdk.LoggingSetup;
import com.cc0mon.sdk.Models.SpeciesImage;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.logging.Logger;

class GetRegistryImages {
    public static void main(String[] args) {
        Logger log = LoggingSetup.configure("get-registry-images");
        Integer limit = null;
        String nameContains = null;
        boolean hasImage = false;
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--limit": if (i + 1 < args.length) try { limit = Integer.parseInt(args[++i]); } catch (NumberFormatException e) {} break;
                case "--name-contains": if (i + 1 < args.length) nameContains = args[++i]; break;
                case "--has-image": hasImage = true; break;
            }
        }
        if (limit != null && limit < 0) {
            System.err.println("validation error: --limit must be >= 0, got " + limit);
            System.exit(5);
        }
        log.info("script start limit=" + limit + " name_contains=" + nameContains + " has_image=" + hasImage);

        try (Client client = new Client()) {
            List<SpeciesImage> images = client.getRegistryImages();
            List<SpeciesImage> filtered = new ArrayList<>(images);
            if (hasImage) filtered.removeIf(i -> i.tokenId() == null);
            if (nameContains != null) {
                String needle = nameContains.toLowerCase(Locale.ROOT);
                filtered.removeIf(i -> {
                    String name = i.name() == null ? "" : i.name().toLowerCase(Locale.ROOT);
                    return !name.contains(needle);
                });
            }
            List<SpeciesImage> shown = limit == null ? filtered : filtered.subList(0, Math.min(limit, filtered.size()));
            System.out.println(new ObjectMapper().enable(SerializationFeature.INDENT_OUTPUT).writeValueAsString(shown));
            log.info("script exit code=0 total=" + images.size() + " matched=" + filtered.size() + " shown=" + shown.size());
        } catch (Errors.ValidationException e) { log.severe("validation: " + e.getMessage()); System.err.println("validation error: " + e.getMessage()); System.exit(5);
        } catch (Errors.NetworkException e)    { log.severe("network: " + e.getMessage());    System.err.println("network error: " + e.getMessage()); System.exit(2);
        } catch (Errors.RateLimitException e)  { log.severe("rate limited http " + e.getStatusCode()); System.err.println("rate limited HTTP " + e.getStatusCode() + ": " + e.getBody()); System.exit(3);
        } catch (Errors.ClientApiException e)  { log.severe("http " + e.getStatusCode());     System.err.println("client error HTTP " + e.getStatusCode() + ": " + e.getBody()); System.exit(3);
        } catch (Errors.ServerApiException e)  { log.severe("http " + e.getStatusCode());     System.err.println("server error HTTP " + e.getStatusCode() + ": " + e.getBody()); System.exit(4);
        } catch (Exception e) { log.severe("unexpected: " + e); System.err.println("unexpected: " + e); System.exit(1); }
    }
}

File: python/examples/cc0mon-api-get-registry-images.py
"""cc0mon-api-get-registry-images — list and filter species image URLs.

Usage::

    python cc0mon-api-get-registry-images.py
    python cc0mon-api-get-registry-images.py --limit 10
    python cc0mon-api-get-registry-images.py --name-contains stryx
    python cc0mon-api-get-registry-images.py --has-image
"""
from __future__ import annotations

import argparse
import dataclasses
import json
import sys

from cc0mon_sdk import (
    Client,
    ClientApiError,
    NetworkError,
    RateLimitError,
    ServerApiError,
    ValidationError,
    configure_logger,
)

ACTION = "get-registry-images"


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="List cc0mon species image URLs, optionally filtered.")
    parser.add_argument("--limit", type=int, default=None, help="Show only the first N entries (default: all)")
    parser.add_argument("--name-contains", dest="name_contains", default=None, help="Case-insensitive substring match on species name")
    parser.add_argument("--has-image", dest="has_image", action="store_true", help="Skip species without a mapped token (tokenId=null)")
    args = parser.parse_args(argv)

    logger = configure_logger(ACTION)
    logger.info("script start limit=%s name_contains=%s has_image=%s", args.limit, args.name_contains, args.has_image)

    try:
        with Client() as client:
            images = client.get_registry_images()
    except ValidationError as e:
        logger.error("validation: %s", e)
        print(f"validation error: {e}", file=sys.stderr)
        return 5
    except NetworkError as e:
        logger.error("network: %s", e)
        print(f"network error: {e}", file=sys.stderr)
        return 2
    except RateLimitError as e:
        logger.error("rate limited http %d: %s", e.status_code, e.body[:200])
        print(f"rate limited HTTP {e.status_code}: {e.body[:200]}", file=sys.stderr)
        return 3
    except ClientApiError as e:
        logger.error("http %d: %s", e.status_code, e.body[:200])
        print(f"client error HTTP {e.status_code}: {e.body[:200]}", file=sys.stderr)
        return 3
    except ServerApiError as e:
        logger.error("http %d: %s", e.status_code, e.body[:200])
        print(f"server error HTTP {e.status_code}: {e.body[:200]}", file=sys.stderr)
        return 4

    filtered = images
    if args.has_image:
        filtered = [i for i in filtered if i.token_id is not None]
    if args.name_contains:
        needle = args.name_contains.lower()
        filtered = [i for i in filtered if needle in (i.name or "").lower()]

    shown = filtered[: args.limit] if args.limit else filtered
    print(json.dumps([dataclasses.asdict(i) for i in shown], indent=2, default=str))
    logger.info("script exit code=0 total=%d matched=%d shown=%d", len(images), len(filtered), len(shown))
    return 0


if __name__ == "__main__":
    sys.exit(main())
